### PR TITLE
[SPARK-17091] Add rule to convert IN predicate to equivalent Parquet filter.

### PR DESCRIPTION
**Pull request by ptkool**
## What changes were proposed in this pull request?

Add a new optimizer rule to convert an IN predicate to an equivalent Parquet filter.

## How was this patch tested?

Tested using unit tests, integration tests, and manual tests.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Is it a duplicate of https://github.com/apache/spark/pull/14671? cc @a10y.

**Comment by a10y** [comment]
Indeed it appears to be. The resolution from my previous PR was that per @HyukjinKwon's benchmarks, performing the disjunction in Spark was slightly more performant than pushing it down to Parquet.

I haven't been following Spark closely these past 12 months so things may have changed.

@ptkool did you do any profiling that would lead you to believe pushing the filter to Parquet leads to perf improvements?

**Comment by ptkool** [comment]
@a10y Yes. Please have a look at my comments in https://issues.apache.org/jira/browse/SPARK-21218.

**Comment by HyukjinKwon** [comment]
(I would like to suggest fix the JIRA in the PR title to point out SPARK-17091)

**Comment by rxin** [comment]
Have you done actual benchmarks to validate that this is a perf improvement?

**Comment by ptkool** [comment]
@rxin Yes, I have.

https://issues.apache.org/jira/browse/SPARK-21218?focusedCommentId=16064608&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16064608

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 1]
```diff
@@ -238,6 +238,14 @@ private[parquet] object ParquetFilters {
       case sources.Not(pred) =>
         createFilter(schema, pred).map(FilterApi.not)
 
+      case sources.In(name, values) if canMakeFilterOn(name) =>
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: Always push-down? Should we also consider the number of elements in `values`? What is the performance impact when the number of values is around 10 or more?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 2]
```diff
@@ -238,6 +238,14 @@ private[parquet] object ParquetFilters {
...
+        val conds = values.flatMap(v => makeEq.lift(nameToType(name)).map(_(name, v)))
+        var filter = conds(0)
+        conds.drop(1).foreach { v =>
```
Comment (a10y) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: You can eliminate the var by using reduceLeft

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
Now we will have this for row group filtering in most cases after #15049. I believe it makes sense in this case.

**Comment by HyukjinKwon** [comment]
I believe I need to cc @jiangxb1987 and @viirya too who activiely reviewed my PR.

**Comment by HyukjinKwon** [comment]
cc @liancheng too who I know is insightful in this.

**Comment by jiangxb1987** [comment]
Please rebase this PR to the latest master. Thanks!

**Comment by viirya** [comment]
I guess this is inactive now.

**Comment by a10y** [comment]
@ptkool are you still tracking this at all?

**Comment by ptkool** [comment]
@a10y Yes, I'm still tracking this.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by wangyum** [comment]
@ptkool Are you still working on?

**Comment by HyukjinKwon** [comment]
@wangyum, can you take over this? Seems it's been inactive long time.

**Pull request closed without merging**